Make connection handling more explicit and remove outdated examples

zvyn opened the pull request:
With this patch we can be more certain that the connection is not kept
for longer than the `read_timeout`.

zvyn committed:
Make connection handling more explicit

With this patch we can be more certain that the connection is not kept
for longer than the `read_timeout`.

zvyn committed:
Remove outdated examples

Refer to https://developer.geops.io for up to date documentation.

zvyn committed:
Quick and dirty unittest fix

aheld84 commented on redis_websocket_api/handler.py:
```diff
@@ -150,9 +150,10 @@ async def listen(self):
-            self.websocket.close(), self.read_timeout
-        )
+        if self.websocket.open:
+            await asyncio.wait_for(  # attempt to say goodbye to the client
```
Was not changed in this MR but maybe it is relevant: In the [websockets documentation](https://websockets.readthedocs.io/en/stable/reference/server.html#websockets.server.WebSocketServerProtocol.close), it says that canceling a `close` is discouraged and one should change `close_timeout` if it takes too long. See [here](https://websockets.readthedocs.io/en/stable/reference/common.html#websockets.legacy.protocol.WebSocketCommonProtocol) for how `close_timeout` is interpreted.

(I hope I found the correct `close` method in the docs)

zvyn replied:
Yes you are right, cancelling `close` skips quite a bit of cleanup code. I simplified the code on our end to let `websockets` deal with all that

aheld84 commented on redis_websocket_api/server.py:
```diff
@@ -49,13 +49,25 @@ async def websocket_handler(self, websocket, path):
+        )
         try:
-            await asyncio.wait_for(handler.listen(), self.keep_alive_timeout)
+            await asyncio.wait(
```
With the new variant, exceptions from `asyncio.wait_for(handler.listen(), self.keep_alive_timeout)` (`TimeOutError` and other exceptions) are silently ignored. Before, they were raised. Is that on purpose?

zvyn replied:
Digging down the rabbit hole I found a few more things to improve... `handler_listen_task` is awaited now, also `handler.close` makes sure waiting for a message in `_queue_reader` is cancelled (meaning `handler_listen_task` is typically done once the `finally` below is reached).

aheld84 replied:
Looks good to me now.
The only thing I'm not really sure about is if it can happen that the `asyncio.create_task(websocket.wait_closed())` task can be garbage-collected before it finishes (and if that would harm anyone if it happens). I assume that `asyncio.wait` keeps a reference to all tasks but as soon as it returns, there is no reference to that task anymore but the connection might still be open at that point, so the task might still be running. Maybe it would be more clean to await it (maybe by using the return values from `asyncio.wait`).

zvyn replied:
`wait_closed` is just a synchronization helper, that's why I didn't really care about how it's cleaned up after the `wait`. But I guess it's better to avoid loose tasks all together, so I'll give it a name

aheld84 reviewed (commented):
There are two points that need clarification. Otherwise it looks good.

zvyn committed:
Remove polling for state, cancel reader instead

This deprecates the `read_timeout` argument but does not brake the
existing public API.

Also switched some early Python 3.5 asyncio coding style for Python 3.7+
patterns. Most notable the `loop` is only used directly if given as
parameter.

aheld84 commented on redis_websocket_api/handler.py:
```diff
@@ -15,18 +17,26 @@
     def __init__(self, redis, websocket, read_timeout=None):
+        if read_timeout:
+            logger.warning(
+                "read_timeout is not used anymore because cleanup is trigered "
+                "immidiatly on connection loss"
```
```suggestion
                "read_timeout is not used anymore because cleanup is triggered "
                "immediately on connection loss"
```

aheld84 commented on redis_websocket_api/handler.py:
```diff
@@ -131,35 +141,41 @@ def channel_is_allowed(self, channel_name):
+
+        if read_timeout:
+            logger.warning(
+                "read_timeout is not used anymore because cleanup is trigered "
+                "immidiatly on connection loss"
```
```suggestion
                "read_timeout is not used anymore because cleanup is triggered "
                "immediately on connection loss"
```

aheld84 commented on redis_websocket_api/server.py:
```diff
@@ -26,13 +26,16 @@ def __init__(
-        self.read_timeout = read_timeout
+        if read_timeout:
+            logger.warning(
+                "read_timeout is not used anymore because cleanup is trigered "
+                "immidiatly on connection loss"
```
```suggestion
                "read_timeout is not used anymore because cleanup is triggered "
                "immediately on connection loss"
```

aheld84 reviewed (commented):
I have one question and some suggestions for typos. Otherwise it looks good.

zvyn committed:
Fix typos in warnings

Co-authored-by: Alexander Held <aheld84@users.noreply.github.com>

zvyn committed:
Await everything regadless of relevance

aheld84 reviewed (approved):
Ready to be merged.
